base/dataset_prepare.go

```diff
@@ -145,46 +145,66 @@ func InferValues(pro *profile.Profile, ds *dataset.Dataset) error {
 	// if we don't have a structure or schema then attempt to determine one
 	body := ds.BodyFile()
 	if body != nil && (ds.Structure == nil || ds.Structure.Schema == nil) {
-		// use a TeeReader that writes to a buffer to preserve data
-		buf := &bytes.Buffer{}
-		tr := io.TeeReader(body, buf)
-		var df dataset.DataFormat
-
-		df, err := detect.ExtensionDataFormat(body.FileName())
-		if err != nil {
-			log.Debug(err.Error())
-			err = fmt.Errorf("invalid data format: %s", err.Error())
+		if err := InferStructure(ds); err != nil {
 			return err
 		}
+	}
 
-		guessedStructure, _, err := detect.FromReader(df, tr)
-		if err != nil {
-			log.Debug(err.Error())
-			err = fmt.Errorf("determining dataset structure: %s", err.Error())
-			return err
-		}
+	if ds.Transform != nil && ds.Transform.ScriptFile() == nil && ds.Transform.IsEmpty() {
+		ds.Transform = nil
+	}
 
-		// attach the structure, schema, and formatConfig, as appropriate
-		if ds.Structure == nil {
-			ds.Structure = guessedStructure
-		}
-		if ds.Structure.Schema == nil {
-			ds.Structure.Schema = guessedStructure.Schema
-		}
-		if ds.Structure.FormatConfig == nil {
-			ds.Structure.FormatConfig = guessedStructure.FormatConfig
-		}
+	return nil
+}
 
-		// glue whatever we just read back onto the reader
-		// TODO (b5)- this may ruin readers that transparently depend on a read-closer
-		// we should consider a method on qfs.File that allows this non-destructive read pattern
-		ds.SetBodyFile(qfs.NewMemfileReader(body.FileName(), io.MultiReader(buf, body)))
+// InferStructure infers the Structure field of the dataset, guarenteeing
+// that the structure will contain a Format, FormatConfig, and Schema, based
+// on the given dataset body. It will not write over any Structure, Format,
+// FormatConfig, or Schema that already exists.
+func InferStructure(ds *dataset.Dataset) error {
+	if ds == nil {
+		return fmt.Errorf("empty dataset")
 	}
 
-	if ds.Transform != nil && ds.Transform.ScriptFile() == nil && ds.Transform.IsEmpty() {
-		ds.Transform = nil
+	body := ds.BodyFile()
+	if body == nil {
+		return fmt.Errorf("empty body")
+	}
+	// use a TeeReader that writes to a buffer to preserve data
+	buf := &bytes.Buffer{}
+	tr := io.TeeReader(body, buf)
+	var df dataset.DataFormat
+
+	df, err := detect.ExtensionDataFormat(body.FileName())
+	if err != nil {
+		log.Debug(err.Error())
+		return fmt.Errorf("invalid data format: %s", err.Error())
+	}
+
+	guessedStructure, _, err := detect.FromReader(df, tr)
```

Comment:
I realize this is pretty much a pure extraction refactor, but just wanted to leave a note for myself. Our usage of `detect` is getting more and more coupled to qri core, and there's evidence (such as ignoring the middle return value) that more changes should be made to it to bring it more in line with how qri specifically uses it. So, I'd like to bring up https://github.com/qri-io/qri/issues/1371 again, which proposes that `detect` move out of the `dataset` repo and into `base/detect` so that it's easier to work with.